### PR TITLE
Datatable selection

### PR DESCRIPTION
**Pull request by ChristianBeilschmidt**

### CODE REVIEW

**Comment by ChristianBeilschmidt** [comment]
[Select Data Tables from Layer Menu](https://trello.com/c/lDnwAB3A/448-select-data-tables-from-layer-menu)

File: projects/wave-core-new/src/lib/tab-panel/tab-panel.component.ts [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+import {LayoutService} from '../layout.service';
+import {TabContent, TabPanelService} from './tab-panel.service';
+
+@Component({
+    selector: 'wave-tab-panel',
+    templateUrl: './tab-panel.component.html',
+    styleUrls: ['./tab-panel.component.scss'],
+    changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class TabPanelComponent implements OnInit, OnChanges, OnDestroy, AfterViewInit {
```
Comment (michaelmattig) [inline projects/wave-core-new/src/lib/tab-panel/tab-panel.component.ts:28]: maybe just call it TabComponent? A Tab panel is only the stuff corresponding to a tab https://www.w3.org/wiki/TabPanel

**Review by michaelmattig — CHANGES_REQUESTED** [review]
some comments c.f. above

File: projects/wave-core-new/src/lib/tab-panel/tab-panel.component.ts (hunk 1 above)
Comment (ChristianBeilschmidt) [reply projects/wave-core-new/src/lib/tab-panel/tab-panel.component.ts]: d280776

**Comment by ChristianBeilschmidt** [comment]
> Some observations where we should discuss whether they make sense and if they are maybe for later
> 
>     * You can add the same layer multiple times to the table, is that intended?
It was not intended to be forbidden. I guess we should discuss what behavior we want to have.

> 
>     * when you close a tab and others are open, the next one should be opened automatically
Which is the next one? The one left to it or the furthermost right one?

> 
>     * wouldn't it make sense to make the table-panel only as tall as the rows are? otherwise it looks like the table has only e.g. 5 elements if you don't look at the buttons on the bottom right because there is so much empty space. On the other hand if I hav 20 elements I have a scroll bar AND a pagination. not user friendly

I guess the component should have a size that is independent of the content. Like in vscode, where your bottom bar doesn't resize on every tab you select.

> 
>     * if I select a tab and change the number of elements per page, change the tab and go back the number of elements per rows is resetted

Hmm, you mean I would have to implement a state management for tabs? The components are destroyed when they close and re-opened. Currently, they only have their input. But this would mean, they are no longer independent of our tab management, i.e. no standalone usable component. I guess we should discuss this also.

> 
>     * for raster layers you could hide the pagination

We should think about what we want to show there and then implement it. I guess it doesn't make sense to spend effort in an unimplemented part.

> 
>     * the close-button could be right-aligned


065cc3a 

> 
>     * would be great to also have the ability to select a layer from the table-panel. Maybe a "new tab button" which shows a layer selection

I don't get it from this description. We should discuss this as well.

**Comment by michaelmattig** [comment]
> 
> 
> > Some observations where we should discuss whether they make sense and if they are maybe for later
> > ```
> > * You can add the same layer multiple times to the table, is that intended?
> > ```
> 
> It was not intended to be forbidden. I guess we should discuss what behavior we want to have.

what would be the use case for opening a data table multiple times? I can't think of one so if a user clicks "show table" and a tab already exists, it should be opened imho.

> > ```
> > * when you close a tab and others are open, the next one should be opened automatically
> > ```
> 
> Which is the next one? The one left to it or the furthermost right one?

the right one like in web browser

> > ```
> > * wouldn't it make sense to make the table-panel only as tall as the rows are? otherwise it looks like the table has only e.g. 5 elements if you don't look at the buttons on the bottom right because there is so much empty space. On the other hand if I hav 20 elements I have a scroll bar AND a pagination. not user friendly
> > ```
> 
> I guess the component should have a size that is independent of the content. Like in vscode, where your bottom bar doesn't resize on every tab you select.

yeah, but why would you want different number of rows in different tabs?

> > ```
> > * if I select a tab and change the number of elements per page, change the tab and go back the number of elements per rows is resetted
> > ```
> 
> Hmm, you mean I would have to implement a state management for tabs? The components are destroyed when they close and re-opened. Currently, they only have their input. But this would mean, they are no longer independent of our tab management, i.e. no standalone usable component. I guess we should discuss this also.

I think this one we absolutely need. you want to scroll to some row in one table, go to a different tab to compare data and go back. Can be a new task but I think it's crucial.


Also a new point: 
- Renaming a layer in the layer list should also rename the tab.
- Layer names, like in the layer list, should be truncated s.t. the tabs are of same width. Right now a very long layer name causes a very wide tab

**Comment by ChristianBeilschmidt** [comment]
> > > Some observations where we should discuss whether they make sense and if they are maybe for later
> > > ```
> > > * You can add the same layer multiple times to the table, is that intended?
> > > ```
> > 
> > 
> > It was not intended to be forbidden. I guess we should discuss what behavior we want to have.
> 
> what would be the use case for opening a data table multiple times? I can't think of one so if a user clicks "show table" and a tab already exists, it should be opened imho.

7879f90 

> 
> > > ```
> > > * when you close a tab and others are open, the next one should be opened automatically
> > > ```
> > 
> > 
> > Which is the next one? The one left to it or the furthermost right one?
> 
> the right one like in web browser
> 


5cc6b8d 

> > > ```
> > > * wouldn't it make sense to make the table-panel only as tall as the rows are? otherwise it looks like the table has only e.g. 5 elements if you don't look at the buttons on the bottom right because there is so much empty space. On the other hand if I hav 20 elements I have a scroll bar AND a pagination. not user friendly
> > > ```
> > 
> > 
> > I guess the component should have a size that is independent of the content. Like in vscode, where your bottom bar doesn't resize on every tab you select.
> 
> yeah, but why would you want different number of rows in different tabs?
> 

I don't get how your point I guess.

> > > ```
> > > * if I select a tab and change the number of elements per page, change the tab and go back the number of elements per rows is resetted
> > > ```
> > 
> > 
> > Hmm, you mean I would have to implement a state management for tabs? The components are destroyed when they close and re-opened. Currently, they only have their input. But this would mean, they are no longer independent of our tab management, i.e. no standalone usable component. I guess we should discuss this also.
> 
> I think this one we absolutely need. you want to scroll to some row in one table, go to a different tab to compare data and go back. Can be a new task but I think it's crucial.
> 

We need a concept for that. First of all, we already have a problem when we move the map 1px. Second, do we want to solve it by injecting specific state tokens to the component (which may interfere in using it as a normal component in other apps)? Or do we want specific services for specific things we display as tab content, e.g. a service that stores the number of rows to display at a time for each table. Third, we should combine this with the concept for selecting rows/things on the map.

> Also a new point:
> 
>     * Renaming a layer in the layer list should also rename the tab.


5c3d8ef 

> 
>     * Layer names, like in the layer list, should be truncated s.t. the tabs are of same width. Right now a very long layer name causes a very wide tab


a6bcf6e

**Comment by michaelmattig** [comment]
grafik
Tabs with long layer names are still wider than regular ones

Also a new issue: imho tabs should be removed if the corresponding layer is removed.

**Comment by ChristianBeilschmidt** [comment]
> grafik
> Tabs with long layer names are still wider than regular ones

a3ccdce 

> 
> Also a new issue: imho tabs should be removed if the corresponding layer is removed.


ce924e2

**Comment by michaelmattig** [comment]
looks good. please resolve conflicts and merge

**Pull request merged**